### DISCUSSION TITLE
UnhandledPromiseRejectionWarning: Error: unable to verify the first certificate

### DISCUSSION THREAD

**Discussion by eliassal** [rfennell/AzurePipelines#1373] [Q&A]
I came across your extension and tried it in Azure Devops server pipline, I get the following error and not sure which certificate it searches

![image](https://user-images.githubusercontent.com/8447864/200356537-4f13bebd-257c-4c31-9fce-0ca8f7df86b4.png)

**Comment by rfennell** [comment]
Is there a proxy or firewall between the agent and the server?

**Comment by eliassal** [comment]
Hello, absolutly no proxy or firewall. It is a local azure devops server in my lan and build agent is in local machine in the same lan and IP, I run pipelines calling sonarqube on remote linux adresse class, nexus or deploy to azure cloud without any issue

**Reply by rfennell** [reply]
There is a report of this being a bad agent install [see this post](https://learn.microsoft.com/en-us/answers/questions/627234/azure-devops-build-nuget-restore-unable-to-verify.html). But given everything else seems OK, it would appear unlikely.

What is the URL of the server? Is it using a self signed cert for HTTPS - [as this report mentions](https://stackoverflow.com/questions/56855546/azure-devops-server-pipeline-build-fails-when-using-self-signed-ssl-certificate). When I last had this issue report #842 this was the problem

**Comment by eliassal** [comment]
It a certificate generated by XCA
![image](https://user-images.githubusercontent.com/8447864/200654254-b7340ba8-017c-457f-ae87-d3cac5f6340b.png)
URL is https://azuredevops2k19.salam.net/

I dont think the issue at https://stackoverflow.com/questions/56855546/azure-devops-server-pipeline-build-fails-when-using-self-signed-ssl-certificate is relevant as my build agent acceses Azure and communications done over https and all build pipelines go through

After reading the 2nd #842, I downloaded the certificate and added it to the computer certificate store but it did not help, is this what is meant by "Problem solved, my self signed certificate was only in the IIS storage but not in the **windows storage**" or it is something else?

**Reply by rfennell** [reply]
Does the setting I mentioned in the issue help?

Or can you try using a connection tonight on http  even for just a test

**Reply by eliassal** [reply]
Yes, adding NODE_TLS_REJECT_UNAUTHORIZED (in the variables section and not in the task section) with a value 0 made it work

```
variables:
  buildConfiguration: 'Release'
   alertSonarIssues: 'nothin'
  stopBuildAccordingToSnarIssues: 0
  stopOrContinue: 0
  dateAndTimeBuildTriggered: ''
  versionProject: ''
  NODE_TLS_REJECT_UNAUTHORIZED: 0
```

The other solution which suggested to add a variable with path to the certificate did not work

> NODE_EXTRA_CA_CERTS: C:\mycertificates\AzurDevopscertificate.crt

**Reply by eliassal** [reply]
but can this be fixed in order to be sure we are on the safe side?

**Reply by eliassal** [reply]
Another question please, once report is generated, how can be pusblished to Azure Devops wiki? Thanks again

**Reply by bm-fez** [reply]
> but can this be fixed in order to be sure we are on the safe side?

There is nothing I can do in my task, it is due to the way the Node runtime handle certs

**Reply by bm-fez** [reply]
> Another question please, once report is generated, how can be pusblished to Azure Devops wiki? Thanks again

Yes, I have another task for that [WIKIUpdater](https://marketplace.visualstudio.com/items?itemName=richardfennellBM.BM-VSTS-WIKIUpdater-Tasks). Have a look at this [conference session](https://www.youtube.com/watch?v=xaV3dFoQdV8&t=332s) I did that show the use of both tasks

**Reply by eliassal** [reply]
Fantastic, it worked on the 2nd try like a  charm as the 1st run I got 

> ##[error]Error: Cloning into 'C:\Projects\agent-AzureDevOps2K20Update15\_work\17\s\repo'...
> fatal: Authentication failed for 'https://azuredevops2k19.salam.net/Sierac-Utilities/PluralSight/_git/PluralSight.wiki/'

as I checked "Run with Build Credentials" (my Build agent uses a PAT which has all rights), I changed it to use my login and passwordannd went through

Why is this if you allow me to ask the question?

Your presentation is also very useful, I watched it all, as I am new to your extensions, in some time, you indicate that we can write our own functions, where we can write them? also the notion of handbalers is a little bit confusing for me? Are they templates?

**Reply by eliassal** [reply]
I also tried to name the file as follows

> Releasenotes-$(Rev:.r).md

but got the following error n the steps

##[error]Error: fatal: pathspec 'Releasenotes-$(Rev:.r).md' did not match any files

Should filename and sourceFile be the same as or not necessary

> filename: 'Releasenotes-$(Rev:.r).md'
> replaceFile: true
> dataIsFile: true
> sourceFile: '$(Build.ArtifactStagingDirectory)/releasenotes.md'
>

**Reply by rfennell** [reply]
> Fantastic, it worked on the 2nd try like a charm as the 1st run I got
> 
> > ##[error]Error: Cloning into 'C:\Projects\agent-AzureDevOps2K20Update15_work\17\s\repo'...
> > fatal: Authentication failed for 'https://azuredevops2k19.salam.net/Sierac-Utilities/PluralSight/_git/PluralSight.wiki/'
> 
> as I checked "Run with Build Credentials" (my Build agent uses a PAT which has all rights), I changed it to use my login and passwordannd went through
> 
> Why is this if you allow me to ask the question?
> 


This error is usually down to the build agent process not have the corrects settings/permissions - they are a bit fiddly to set up [see this repos wiki](https://github.com/rfennell/AzurePipelines/wiki/WIKI-Updater-Tasks#authentication-using-oauth-to-azure-devops-services-hosted-repos)

**Reply by rfennell** [reply]
> Your presentation is also very useful, I watched it all, as I am new to your extensions, in some time, you indicate that we can write our own functions, where we can write them? also the notion of handbalers is a little bit confusing for me? Are they templates?

The Handlebars template you provide defines the layout of the release notes. Handlebars.JS has a number of [built in functions](https://handlebarsjs.com/guide/) and I also include [a large library of common functions ](https://github.com/helpers/handlebars-helpers) to help parse/process data.

If none of these do the job you required (which is fairly rare)  you have the option of writing your own Handlebar function (in JavaScript). 

You can find sample templates and functions in this [repo folder](https://github.com/rfennell/AzurePipelines/tree/main/SampleTemplates/XplatGenerateReleaseNotes%20(Node%20based)/Version%203)

**Reply by rfennell** [reply]
> ##[error]Error: fatal: pathspec 'Releasenotes-$(Rev:.r).md' did not match any files

You cannot use the `$(Rev:.r)` syntax anywhere other than when defining the build number. For task parameters you have to use either [Azure DevOps predefined variables](https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml) or user defined ones.

**Reply by eliassal** [reply]
OK but i need to have a page per build/Release number so that people know to which build is related the wiki page, isnt logic?Very strange from MS to prevent using the revision as a variable!!

**Reply by rfennell** [reply]
> OK but i need to have a page per build/Release number so that people know to which build is related the wiki page, isnt logic?Very strange from MS to prevent using the revision as a variable!!

That is easy to solve.

In your YAML set the build name, using any mixture of user defined variables and the special build number options e.g.

`name: ' $(major).$(minor).$(year:yy)$(dayofyear).$(rev:r)'`

Of course this can also be set in the UX if you are using Classic Builds

![image](https://user-images.githubusercontent.com/324718/200874100-825b9c86-62d2-43b3-a010-abfc37bd5047.png)

This build number is now available via the pre-defined variable `$(Build.BuildNumber)` e.g.

        - task: WikiUpdaterTask@1
          displayName: 'Update Project Wiki with release notes'
          inputs:
            repo: '$(WikiRepoUrl)'
            filename: 'Builds/`Notes $(Build.BuildNumber).md'
            replaceFile: True
            sourceFile: '$(Build.ArtifactStagingDirectory)\Changes_$(GitVersion.FullSemVer).md'
            dataIsFile: True
            message: 'Added release notes for build $(Build.BuildNumber)'
            localpath: '$(System.DefaultWorkingDirectory)\repo'
            useAgentToken: true

**Reply by eliassal** [reply]
Hello Richard, the statement

> ' $(major).$(minor).$(year:yy)$(dayofyear).$(rev:r)'

 does not give what I expect, it retrn

> $(major).$(minor).22313.2

I did not want to bother you yesterday and opened a ticket with ADOS community but did not get an appropriate response

**Reply by rfennell** [reply]
`$(major)` and `$(minor)` are user defined variable i.e. you have to add them to your pipeline and give them values. if you don't do this then the text is just rendered as you are seeing - I only mentioned them as an example of what could be done.

The format of your build name is down to you, here are no end of options

- `name: ' $(major).$(minor).$(year:yy)$(dayofyear).$(rev:r)'` - using a mix of user defined and system variables
- `name: ' 1.2.$(year:yy)$(dayofyear).$(rev:r)'` - using a mix hard coded number and system variables
- `name: ' $(year:yy)$(dayofyear).$(rev:r)'` - using just and system variables

And that is before you star looking at tools like [GitVersion](https://marketplace.visualstudio.com/items?itemName=gittools.gittools) to automatically generate the version numbers

**Reply by eliassal** [reply]
So many thanks, I thought they were system predefined variables. OK I will check GitVersion right now

**Reply by rfennell** [reply]
We use GitVersion for most projects these days, so a YAML is something like

```yaml
name: $(Build.DefinitionName)_$(GitVersion.FullSemVer)

stages:
  - stage: 'Build_Packages'
    jobs:
      - job: 'Build_and_Test'
        pool:
          vmImage: 'windows-latest'
        steps:
        - checkout: self
        # GitVersion task is needed in each job where the variables are referenced
        - task: gitversion/setup@0
          displayName: 'Get current version of GitVersion'
          inputs:
            versionSpec: '5.x'

        - task: gitversion/execute@0
          displayName: 'Run GitVersion to generate SEMVER'
          inputs:
            # only need a config file for some special cases, not essential
            useConfigFile: true
            configFilePath: '$(System.DefaultWorkingDirectory)/GitVersion.yml'

   --- more steps
```

Note
-  that the build name is only set after the pair of GitVersion tasks have run
- You can also run [GitVersion locally on your dev machine](https://gitversion.net/docs/usage/cli/installation) to help work out how it works. The task just wrapper this command

**Reply by eliassal** [reply]
Thanks again for your help, tell me GitVersion.yml  should be there before we run the task or it will be created by GitVersion?

**Reply by rfennell** [reply]
You don't need a gitversion.yml unless you need to change the default behaviour of GitVersion.

Reasons you might have one is
- You want to start numbering at a place other than 0.0.0
- You want to force incrementing the version other than based on commit comments and branch merges
- Your branch names are not the defaults (it used to be you needed a gitversion.yml if the trunk branch was not called master so our file used to need to be as follows to support either master or main) e.g

```
branches:
  master:
    regex: ^master$|^main$
```

For all the options see https://gitversion.net/docs/reference/configuration

**Comment by eliassal** [comment]
I enabled system.debug, here are copie of lats lines

##[debug]checkStage=false
##[debug]getAllParents=False
##[debug]tags=undefined
##[debug]overridePat=undefined
##[debug]overrideBuildReleaseId=undefined
##[debug]getIndirectPullRequests=False
##[debug]stopOnError=False
##[debug]considerPartiallySuccessfulReleases=False
##[debug]checkForManuallyLinkedWI=False
##[debug]wiqlWhereClause=undefined
##[debug]getPRDetails=true
##[debug]getTestedBy=true
##[debug]wiqlFromTarget=WorkItems
##[debug]maxRetries=20
##[debug]Build.BuildId=495
##[debug]Release.ReleaseId=undefined
##[debug]Release.DefinitionId=undefined
##[debug]Release_EnvironmentName=undefined
##[debug]ReleaseNotes.Fix349=undefined
Creating Azure DevOps API connections for https://azuredevops2k19.salam.net/Sierac-Utilities/ with 'allowRetries' set to 'true' and 'maxRetries' count to '20'
##[debug]Getting System.AccessToken
##[debug]Getting credentials the agent is running as
##[debug]SYSTEMVSSCONNECTION exists true
##[debug]Found an OAUTH token
Creating the credential handler from the OAUTH token
(node:2104) UnhandledPromiseRejectionWarning: Error: unable to verify the first certificate
    at TLSSocket.onConnectSecure (_tls_wrap.js:1088:34)
    at TLSSocket.emit (events.js:198:13)
    at TLSSocket._finishInit (_tls_wrap.js:666:8)
(node:2104) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:2104) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Finishing: Generate Release Notes based on Release Comparison API

**Reply by rfennell** [reply]
Can you try setting this variable as detailed in #842 and #801, when this issue has appeared before it was down to how Node handles certs

> This is I think a duplicate of #801. It seems to be down to the use of self signed certificate 
> 
> Anyway I added some steps to try at the end of #801 i.e Azure DevOps variable NODE_TLS_REJECT_UNAUTHORIZED=0 - have a read of that issue and give the steps a try to see if they help
